cub/device/dispatch/dispatch_batch_memcpy.cuh

```diff
@@ -0,0 +1,690 @@
+  __global__ void MultiBlockBatchMemcpyKernel(InputBufferIt input_buffer_it,
+                                              OutputBufferIt output_buffer_it,
+                                              BufferSizeIteratorT buffer_sizes,
+                                              BufferTileOffsetItT buffer_tile_offsets,
+                                              TileT buffer_offset_tile,
+                                              TileOffsetT last_tile_offset)
+{
+  using StatusWord    = typename TileT::StatusWord;
+  using ActivePolicyT = typename ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT;
+  using BufferSizeT   = typename std::iterator_traits<BufferSizeIteratorT>::value_type;
+
+  constexpr uint32_t BLOCK_THREADS    = ActivePolicyT::BLOCK_THREADS;
+  constexpr uint32_t ITEMS_PER_THREAD = ActivePolicyT::BYTES_PER_THREAD;
+  constexpr uint32_t ITEMS_PER_BLOCK  = BLOCK_THREADS * ITEMS_PER_THREAD;
+
+  BufferOffsetT num_blev_buffers = buffer_offset_tile.LoadValid(last_tile_offset);
+
+  uint32_t tile_id = blockIdx.x;
+
+  // No block-level buffers => we're done here
+  if (num_blev_buffers == 0)
+    return;
+
+  // While there's still tiles of bytes from block-level buffers to copied
+  do
+  {
+    __shared__ BufferOffsetT buffer_id;
+
+    // Make sure thread 0 does not overwrite the buffer id before other threads have finished with
+    // the prior iteration of the loop
+    CTA_SYNC();
+
+    // Binary search the buffer that this tile belongs to
+    if (threadIdx.x == 0)
+      buffer_id = UpperBound(buffer_tile_offsets, num_blev_buffers, tile_id) - 1;
+
+    // Make sure thread 0 has written the buffer this thread block is assigned to
+    CTA_SYNC();
+
+    // The relative offset of this tile within the buffer it's assigned to
+    BufferSizeT tile_offset_within_buffer =
+      static_cast<BufferSizeT>(tile_id - buffer_tile_offsets[buffer_id]) * ITEMS_PER_BLOCK;
+
+    // If the tile has already reached beyond the work of the end of the last buffer
+    if (buffer_id >= num_blev_buffers - 1 && tile_offset_within_buffer > buffer_sizes[buffer_id])
+    {
+      return;
+    }
+
+    // Tiny remainders are copied without vectorizing laods
+    if (buffer_sizes[buffer_id] - tile_offset_within_buffer <= 32)
+    {
+      BufferSizeT thread_offset = tile_offset_within_buffer + threadIdx.x;
+      for (int i = 0; i < ITEMS_PER_THREAD; i++)
+      {
+        if (thread_offset < buffer_sizes[buffer_id])
+        {
+          reinterpret_cast<uint8_t *>(output_buffer_it[buffer_id])[thread_offset] =
+            reinterpret_cast<const uint8_t *>(input_buffer_it[buffer_id])[thread_offset];
+        }
+        thread_offset += BLOCK_THREADS;
+      }
+    }
+    else
+    {
+      VectorizedCopy<BLOCK_THREADS, uint4>(
+        threadIdx.x,
+        &reinterpret_cast<char *>(output_buffer_it[buffer_id])[tile_offset_within_buffer],
+        min(buffer_sizes[buffer_id] - tile_offset_within_buffer, ITEMS_PER_BLOCK),
```

Comment:
It seems that `cub::min` expands to using the conditional operator. My past experience was that using conditional is slower than using `min()`. This may be outdated, though. Do you have experience on that? Otherwise, it's probably worth verifying again.